test/metabase/query_processor/middleware/expand_macros_test.clj

```diff
@@ -22,49 +27,80 @@
+                                                  :definition {:filter [:= [:field 5 nil] "abc"]}}
+                                                 {:id         2
+                                                  :name       "Segment 2"
+                                                  :table-id   (meta/id :venues)
+                                                  :definition {:filter [:is-null [:field 7 nil]]}}]})
+                                    meta/metadata-provider)
+    (is (= (mbql-query
             {:filter   [:and
-                        [:segment segment-1-id]
+                        [:= [:field 5 nil] "abc"]
                         [:or
-                         [:segment segment-2-id]
+                         [:is-null [:field 7 nil]]
                          [:> [:field 4 nil] 1]]]
-             :breakout [[:field 17 nil]]}))))))
+             :breakout [[:field 17 nil]]})
+           (#'expand-macros/expand-metrics-and-segments
+            (mbql-query
+             {:filter   [:and
+                         [:segment 1]
+                         [:or
+                          [:segment 2]
+                          [:> [:field 4 nil] 1]]]
+              :breakout [[:field 17 nil]]}))))))
 
-(deftest nested-segments-test
-  (t2.with-temp/with-temp
-    [:model/Segment {s1-id :id} {:table_id   (mt/id :venues)
-                                 :definition {:filter [:< (mt/id :venues :price) 3]}}
-     :model/Segment {s2-id :id} {:table_id   (mt/id :venues)
-                                 :definition {:filter [:and [:segment s1-id] [:> (mt/id :venues :price) 1]]}}]
-    (testing "Nested segments are correctly expanded (#30866)"
-      (is (= (mt/mbql-query venues {:filter [:and [:< $price 3] [:> $price 1]]})
-             (#'expand-macros/expand-metrics-and-segments
-              (mt/mbql-query venues {:filter [:segment s2-id]})))))
+(deftest ^:parallel nested-segments-test
+  (let [metadata-provider (lib/composed-metadata-provider
+                           (lib.tu/mock-metadata-provider
+                            {:segments [{:id         1
+                                         :name       "Segment 1"
+                                         :table-id   (meta/id :venues)
+                                         :definition {:filter [:< [:field (meta/id :venues :price) nil] 3]}}
+                                        {:id         2
+                                         :name       "Segment 2"
+                                         :table-id   (meta/id :venues)
+                                         :definition {:filter [:and
+                                                               [:segment 1]
+                                                               [:> [:field (meta/id :venues :price) nil] 1]]}}]})
+                           meta/metadata-provider)]
+    (qp.store/with-metadata-provider metadata-provider
```

Comment:
This structure appears in a few places, and it seems like an opportunity for a combined macro. Doesn't need to be in this PR.

Reply:
good idea. I can just create one mock metadata provider and share it between a lot of tests